File: backend/app/models/note.py
from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import DateTime, ForeignKey, Index, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class Note(Base):
    __tablename__: str = "notes"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False, default="")
    order_index: Mapped[int] = mapped_column(nullable=False, default=0)
    user_id: Mapped[UUID] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    parent_id: Mapped[UUID | None] = mapped_column(
        ForeignKey("notes.id", ondelete="CASCADE"),
        nullable=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    user: Mapped["User"] = relationship(back_populates="notes")  # pyright: ignore[reportUndefinedVariable]  # noqa: F821
    parent: Mapped["Note | None"] = relationship(remote_side=[id], back_populates="children")
    children: Mapped[list["Note"]] = relationship(back_populates="parent", cascade="all, delete-orphan")

    __table_args__: tuple[Index, Index] = (
        Index("ix_notes_user_id", "user_id"),
        Index("ix_notes_parent_id", "parent_id"),
    )

File: frontend/src/models/note.ts
export interface Note {
    id: string;
    title: string;
    parent_id: string | null;
    content: string;
    order_index: number;
    created_at: string;
    updated_at: string;
}

export interface NoteTreeSelection {
    [key: string]: boolean;
}

export interface NoteInput {
    title: string;
    parent_id: string | null;
    content?: string;
    order_index?: number | null;
}

export interface NoteUpdateInput {
    title?: string;
    parent_id?: string | null;
    content?: string;
    order_index?: number;
}

export interface NoteReorderItem {
    id: string;
    parent_id: string | null;
    order_index: number;
}
